src/shogun/clustering/KMeans.cpp

```diff
@@ -4,446 +4,187 @@
-{
-	if (!R.vector)
-		return SGMatrix<float64_t>();
-
-	CDenseFeatures<float64_t>* lhs=
-		(CDenseFeatures<float64_t>*)distance->get_lhs();
-	SGMatrix<float64_t> centers=lhs->get_feature_matrix();
-	SG_UNREF(lhs);
-	return centers;
-}
-
-int32_t CKMeans::get_dimensions()
-{
-	return dimensions;
-}
-
-void CKMeans::set_fixed_centers(bool fixed)
-{
-	fixed_centers=fixed;
-}
-
-bool CKMeans::get_fixed_centers()
-{
-	return fixed_centers;
-}
-
-void CKMeans::store_model_features()
-{
-	/* set lhs of underlying distance to cluster centers */
-	CDenseFeatures<float64_t>* cluster_centers=new CDenseFeatures<float64_t>(
-			mus);
-
-	/* store cluster centers in lhs of distance variable */
-	CFeatures* rhs=distance->get_rhs();
-	distance->init(cluster_centers, rhs);
-	SG_UNREF(rhs);
-}
-
-SGMatrix<float64_t> CKMeans::kmeanspp()
-{
-	int32_t lhs_size;
-	CDenseFeatures<float64_t>* lhs=(CDenseFeatures<float64_t>*)distance->get_lhs();
-	lhs_size=lhs->get_num_vectors();
-	
-	SGMatrix<float64_t> centers=SGMatrix<float64_t>(dimensions, k);
-	centers.zero();
-	SGVector<float64_t> min_dist=SGVector<float64_t>(lhs_size);
-	min_dist.zero();
-
-	/* First center is chosen at random */
-	int32_t mu=CMath::random((int32_t) 0, lhs_size-1);
-	SGVector<float64_t> mu_first=lhs->get_feature_vector(mu);	
-	for(int32_t j=0; j<dimensions; j++)
-		centers(j, 0)=mu_first[j];
-
-	distance->precompute_lhs();
-	distance->precompute_rhs();
-#pragma omp parallel for shared(min_dist)
-	for(int32_t i=0; i<lhs_size; i++)
-		min_dist[i]=CMath::sq(distance->distance(i, mu));
-#ifdef HAVE_LINALG
-	float64_t sum=linalg::vector_sum(min_dist);
-#else //HAVE_LINALG
-	Map<VectorXd> eigen_min_dist(min_dist.vector, min_dist.vlen);
-	float64_t sum=eigen_min_dist.sum();
-#endif //HAVE_LINALG
-	int32_t n_rands=2 + int32_t(CMath::log(k));
+		/* Update Step : Calculate new means */
+		if (!fixed_centers)
+		{
+			/* mus=zeros(dim, num_centers) ; */
+			centers.zero();
+			for (int32_t i=0; i<lhs_size; i++)
+			{
+				int32_t cluster_i=cluster_assignments[i];
 
-	/* Choose centers with weighted probability */
-	for(int32_t i=1; i<k; i++)
-	{	
-		int32_t best_center=0;		
-		float64_t best_sum=-1.0;
-		SGVector<float64_t> best_min_dist=SGVector<float64_t>(lhs_size);
+				SGVector<float64_t>vec=lhs->get_feature_vector(i);
 
-		/* local tries for best center */
-		for(int32_t trial=0; trial<n_rands; trial++)
-		{
-			float64_t temp_sum=0.0;		
-			float64_t temp_dist=0.0;
-			SGVector<float64_t> temp_min_dist=SGVector<float64_t>(lhs_size);		
-			int32_t new_center=0;		
-			float64_t prob=CMath::random(0.0, 1.0);
-			prob=prob*sum;
+				for (int32_t j=0; j<dim; j++)
+					centers(j, cluster_i) += vec[j];
+				lhs->free_feature_vector(vec, i);
+			}
 		
-			for(int32_t j=0; j<lhs_size; j++)
+			for (int32_t i=0; i<num_centers; i++)
 			{
-				temp_sum+=min_dist[j];
-				if (prob <= temp_sum)
+				if (weights_set[i]!=0.0)
 				{
-					new_center=j;
-					break;
+					for (int32_t j=0; j<dim; j++)
```

Comment:
mmm this is basically a scaling of the centers_i column with weights_set[i];
you should use a vector operation on this.